packages/request-state/src/runtime/middleware.ts

```diff
@@ -0,0 +1,22 @@
+import type { ReadableStream } from 'node:stream/web';
+import { defineMiddleware } from 'astro/middleware';
+import { collectState } from './serverState.js';
+
+export const onRequest = defineMiddleware(async (_, next) => {
+	const { getState, result } = await collectState(next);
+
+	async function* render() {
+		for await (const chunk of result.body as ReadableStream<ArrayBuffer>) {
+			yield chunk;
+		}
+
+		const state = getState();
+
+		if (state) {
+			yield `<script id="it-astro-state" type="application/json+devalue">${state}</script>`;
```

Comment:
How can you know where it will end up in the html (eg. is it better in the head or the body? can it end up above the doctype?)

Reply:
The state is only accessible from a module, which means it will run only after the full HTML is parsed, so this can be anywhere on the document and it will still work fine.

By yielding it at the end it will always be added after the `</html>` closing the page, which is valid for this kind of use.